# RNASeq-Pipeline
Simplified Bioinformatics Pipeline for Scientists Analyzing RNASeq Data

## Table of Contents
- [Setting Up Your Work Environment](#environment)
- [Working in a Cloud Environment](#cloud-computing)
  - [Vanderbilt ACCRE Cluster Cloud Computing](#vanderbilt-accre)
- [Transcript Quantification with Salmon](#salmon)
- [Quality Checking with FastQC Analysis](#fastqc)
- [Differential Gene Expression Analysis](#differential-gene-expression-analysis)
  - [DESeq2](#deseq2)
  - [Plots](#plots)
    - [Principle Component Analysis (PCA)](#principle-component-analysis)
    - [Heat Maps](#heat-maps)
  - [Normalized Counts Data](#normalized-counts)
    - [MA Plots](#ma-plots)
    - [Heat Maps](#heat-maps)
  - [Pairwise Analysis](#pairwise-analysis)
    - [Volcano Plots](#volcano-plots)
  - [Gene Set Enrichment Analysis (GSEA)](gene-set-enrichment-analysis)
- 

## Environment
> How to set up the basic work environment for the RNASeq Pipeline,

1. Clone this repository
Open a terminal window and navigate to the place you would like to download this package using the `cd` command.
Then, clone the repository using one of 2 commands with the link in the green code button at the top left of this repo.
```bash
git clone https://github.com/parkdj1/RNASeq-Pipeline.git    # HTTPS version (login using email/password)
git clone git@github.com:parkdj1/RNASeq-Pipeline.git        # SSH   version (login using an ssh key pair)
```
Alternatively, download the zip file where you want the folder to be and open it to unzip.

2. Python
3. R


## Cloud Computing
> Working in a cloud computing environment 101

### Vanderbilt ACCRE
> Vanderbilt University's Cluster Computing Resource - ACCRE

## Salmon

## FastQC

## Differential Gene Expression Analysis

### DeSeq2

### Plots

#### Principle Component Analysis

#### Heat Maps

### Normalized Counts

#### MA Plots

#### Heat Maps

### Pairwise Analysis

#### Volcano Plots

#### Gene Set Enrichment Analysis
